### PR TITLE
feat!: add one-sided coinbase payments

### PR DESCRIPTION
**Pull request by hansieodendaal**
Description
---
Added normal one-sided and stealth one-sided coinbase transactions. Changes are:
- Coinbases can now only be paid to a nominated wallet address directly. 
- The minotari miner and merge mining proxy will only start if a valid minotari wallet address has been supplied for the particular network.
- The miner and merge mining proxy does not depend on the wallet anymore to construct the coinbase; it is done directly.
- Corresponding gRPC method (`rpc GetCoinbase (GetCoinbaseRequest) returns (GetCoinbaseResponse)`)  has been removed from the wallet.
- A memory-based transactions key manager has been created for use by the miner and merge mining proxy as they do not need persistent storage of keys. This also ensures that new entropy for the generation of keys is created each time the miner or merge mining proxy is started effectively negating any collisions of private keys.
- All unused code in the output manager and transaction manager have been removed.

Motivation and Context
---
See #5930

How Has This Been Tested?
---
Unit tests
Cucumber tests
System-level tests - _Completed_

What process can a PR reviewer use to test or verify this change?
---
- Code walk-through
- Run cucumber `Scenario: Get Transaction Info` and review the log files for `WALLET_A` where the stealth one-sided coinbases will be imported and where payment is made to `WALLET_B`.

Breaking Changes
---

- [ ] None
- [ ] Requires data directory on base node to be deleted
- [ ] Requires hard fork
- [x] Other - Please specify

- Coinbases can now only be paid to a nominated wallet address directly.
- Existing wallets need to be recovered into new wallets.

### CODE REVIEW

**Comment by stringhandler** [comment]
Added exclamation for breaking change

**Review by hansieodendaal — COMMENTED** [review]
System-level testing: Merge mining proxy should panic if the wallet payment address is not set; currently it is logging an error.

**Review by hansieodendaal — COMMENTED** [review]
We need more detail about imported coinbase transactions on the console wallet's  transactions pane

**Review by hansieodendaal — COMMENTED** [review]
Many of the imported coinbase transactions in the nominated wallet have the incorrect status. We expect `FauxUnconfirmed` -> `FauxConfrimed`, but they end up as `FauxUnconfirmed` -> `Completed`. The `Completed` transactions are then broadcast to the mempool, which is just a further symptom to the actual problem.
image
image

File: applications/minotari_console_wallet/README.md [hunk 1]
```diff
@@ -195,13 +195,13 @@ Total value of UTXOs: 36105.165440 T
 example output - `--csv-file` (contents of `utxos.csv`)
 
 ```
-"#","Value (uT)","Spending Key","Commitment","Flags","Maturity"
```
Comment (SWvheerden) [inline applications/minotari_console_wallet/README.md]: Why has this disapeared?

File: applications/minotari_console_wallet/src/grpc/mod.rs [hunk 2]
```diff
@@ -51,10 +49,6 @@ pub fn convert_to_transaction_event(event: String, source: TransactionWrapper) -
             direction: "inbound".to_string(),
             amount: inbound.amount.as_u64(),
             message: inbound.message.clone(),
-            // The coinbase are technically Inbound.
-            // To determine whether a transaction is coinbase
-            // we will check whether the message contains `Coinbase`.
-            is_coinbase: inbound.message.to_lowercase().contains("coinbase"),
```
Comment (SWvheerden) [inline applications/minotari_console_wallet/src/grpc/mod.rs:57]: and here

File: applications/minotari_console_wallet/src/init/mod.rs [hunk 3]
```diff
@@ -188,7 +188,6 @@ fn display_password_feedback(passphrase: &SafePassword) -> bool {
             "The password you chose is weak; a determined attacker with access to your device may be able to guess it."
         );
         println!("You may want to consider changing it to a stronger one.");
-        println!("Here are some suggestions:");
```
Comment (SWvheerden) [inline applications/minotari_console_wallet/src/init/mod.rs]: why remove this?

File: base_layer/core/src/transactions/coinbase_builder.rs [hunk 4]
```diff
@@ -306,10 +370,71 @@ where TKeyManagerInterface: TransactionKeyManagerInterface
     }
 }
 
+pub async fn generate_coinbase(
+    fee: MicroMinotari,
+    reward: MicroMinotari,
+    height: u64,
+    extra: &[u8],
+    key_manager: &MemoryDbKeyManager,
+    miner_node_script_key_id: &TariKeyId,
```
Comment (SWvheerden) [inline base_layer/core/src/transactions/coinbase_builder.rs]: this name is used all over and is super confusing as to what it is. 
This is the script key that the spender needs to know when to spend it. And this is only used in tests where we need the wallet_output. 
I would create two functions, one that creates an walletouput, and one that does not. This means that production code needs to randomly generate a key and through this away. 
This all causes confusing and is going to cause bugs in the future.

File: applications/minotari_miner/src/run_miner.rs [hunk 5]
```diff
@@ -66,10 +75,38 @@ pub async fn start_miner(cli: Cli) -> Result<(), ExitError> {
     let mut config = MinerConfig::load_from(&cfg).expect("Failed to load config");
     debug!(target: LOG_TARGET_FILE, "{:?}", config);
     setup_grpc_config(&mut config);
+    let key_manager = create_memory_db_key_manager();
+    let wallet_private_key = PrivateKey::random(&mut OsRng);
+    let miner_node_script_key_id = key_manager
```
Comment (SWvheerden) [inline applications/minotari_miner/src/run_miner.rs]: this just causes confusion,

File: applications/minotari_merge_mining_proxy/src/block_template_protocol.rs [hunk 6]
```diff
@@ -46,20 +55,31 @@ const LOG_TARGET: &str = "minotari_mm_proxy::proxy::block_template_protocol";
...
         base_node_client: &'a mut BaseNodeClient<InterceptedService<Channel, ClientAuthenticationInterceptor>>,
-        wallet_client: &'a mut WalletClient<InterceptedService<Channel, ClientAuthenticationInterceptor>>,
         config: Arc<MergeMiningProxyConfig>,
-    ) -> Self {
-        Self {
-            base_node_client,
-            wallet_client,
+    ) -> Result<BlockTemplateProtocol<'a>, MmProxyError> {
+        let key_manager = create_memory_db_key_manager();
+        let wallet_private_key = PrivateKey::random(&mut OsRng);
```
Comment (SWvheerden) [inline applications/minotari_merge_mining_proxy/src/block_template_protocol.rs]: why?

File: base_layer/wallet/migrations/2023-11-14-131400_coinbase/up.sql [hunk 7]
```diff
@@ -0,0 +1,69 @@
+-- Any old 'outputs' will not be valid due to the removal of 'coinbase_block_height',
```
Comment (SWvheerden) [inline base_layer/wallet/migrations/2023-11-14-131400_coinbase/up.sql]: why remove this? we can still add this in?

File: base_layer/wallet/migrations/2023-11-14-131400_coinbase/up.sql [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+    mined_timestamp                         DATETIME            NULL,
+    encrypted_data                          BLOB                NOT NULL,
+    minimum_value_promise                   BIGINT              NOT NULL,
+    source                                  INTEGER             NOT NULL DEFAULT 0,
+    last_validation_timestamp               DATETIME            NULL,
+    CONSTRAINT unique_commitment UNIQUE (commitment)
+);
+
+-- Any old 'completed_transactions' will not be valid due to the removal of 'coinbase_block_height',
+-- so we drop and recreate the table.
```
Comment (SWvheerden) [inline base_layer/wallet/migrations/2023-11-14-131400_coinbase/up.sql]: why remove this?

File: base_layer/wallet/src/output_manager_service/storage/models.rs [hunk 9]
```diff
@@ -114,27 +114,48 @@ impl Eq for DbWalletOutput {}
...
-    fn from(value: u32) -> Self {
+// impl From<u32> for SpendingPriority {
+//     fn from(value: u32) -> Self {
+//         match value {
+//             0 => SpendingPriority::Normal,
+//             1 => SpendingPriority::HtlcSpendAsap,
+//             _ => SpendingPriority::Unknown,
+//         }
+//     }
+// }
```
Comment (SWvheerden) [inline base_layer/wallet/src/output_manager_service/storage/models.rs]:
```suggestion
```

File: base_layer/wallet/src/output_manager_service/storage/models.rs [hunk 10]
```diff
@@ -114,27 +114,48 @@ impl Eq for DbWalletOutput {}
 
 // ---------------------------------------------------------------------------
 
-#[derive(Debug, Clone)]
+#[derive(Debug, Clone, Eq, PartialEq)]
 pub enum SpendingPriority {
     Normal,
     HtlcSpendAsap,
-    Unknown,
+    // Unknown,
```
Comment (SWvheerden) [inline base_layer/wallet/src/output_manager_service/storage/models.rs]:
```suggestion
```

File: base_layer/wallet/src/output_manager_service/storage/models.rs [hunk 11]
```diff
@@ -114,27 +114,48 @@ impl Eq for DbWalletOutput {}
...
 
-impl From<SpendingPriority> for u32 {
+// impl From<SpendingPriority> for u32 {
+//     fn from(value: SpendingPriority) -> Self {
+//         match value {
+//             SpendingPriority::Normal => 0,
+//             SpendingPriority::HtlcSpendAsap => 1,
+//         }
+//     }
+// }
```
Comment (SWvheerden) [inline base_layer/wallet/src/output_manager_service/storage/models.rs]:
```suggestion
```

**Review by SWvheerden — COMMENTED** [review]
looking good, a few comments

File: applications/minotari_console_wallet/README.md (hunk 1 above)
Comment (hansieodendaal) [reply applications/minotari_console_wallet/README.md:198]: The flags were removed, so it changed to follow suit

Comment (hansieodendaal) [reply applications/minotari_console_wallet/README.md]: Fixed

File: applications/minotari_console_wallet/src/grpc/mod.rs (hunk 2 above)
Comment (hansieodendaal) [reply applications/minotari_console_wallet/src/grpc/mod.rs:57]: We do not have coinbase transactions anymore; we just import one-sided coinbase UTXOs and make faux transactions

File: applications/minotari_console_wallet/src/init/mod.rs (hunk 3 above)
Comment (hansieodendaal) [reply applications/minotari_console_wallet/src/init/mod.rs]: Oops, finger trouble.

File: base_layer/wallet/migrations/2023-11-14-131400_coinbase/up.sql (hunk 7 above)
Comment (hansieodendaal) [reply base_layer/wallet/migrations/2023-11-14-131400_coinbase/up.sql]: The `coinbase_block_height`  column was only used to manage coinbase transactions in the transaction service; this is not used anymore due to the removal of conbase transactions.

File: base_layer/wallet/migrations/2023-11-14-131400_coinbase/up.sql (hunk 8 above)
Comment (hansieodendaal) [reply base_layer/wallet/migrations/2023-11-14-131400_coinbase/up.sql]: See above

File: applications/minotari_merge_mining_proxy/src/block_template_protocol.rs (hunk 6 above)
Comment (hansieodendaal) [reply applications/minotari_merge_mining_proxy/src/block_template_protocol.rs]: removed

**Review by SWvheerden — APPROVED** [review]

**Pull request merged**